user/how-to-guides/backup-emergency-restore-v4.md

```diff
@@ -128,36 +117,53 @@ any GNU/Linux system.
         hmac-algorithm=scrypt
         backup-id=20161020T123455-1234
 
- 6. Set `backup_id` to the value in the last line of `backup-header`. (Note that
+ 5. Set `backup_id` to the value in the last line of `backup-header`. (Note that
     there is a hyphen in `backup-id` in the file, whereas there is an underscore
     in `backup_id` in the variable you're setting.)
 
         [user@restore ~]$ backup_id=20161020T123455-1234
 
- 7. Choose a qube whose data you wish to restore (in this example, `vm1`).
-    Verify the data's integrity, decrypt it, decompress it, and extract it.
+ 6. Verify and decrypt, decompress, and extract the `qubes.xml` file.
 
-        [user@restore ~]$ find vm1 -name 'private.img.*.enc' | sort -V | while read f_enc; do \
-            f_dec=${f_enc%.enc}; \
-            echo "$backup_id!$f_dec!$backup_pass" | scrypt dec -P $f_enc || break; \
-            done | gzip -d | tar -xv
-        vm1/private.img
+        [user@restore ~]$ echo "$backup_id!qubes.xml.000!$backup_pass" |\
+            scrypt dec -P qubes.xml.000.enc | gzip -d | tar -xv
+        qubes.xml
 
     If this pipeline fails, it is likely that the backup is corrupted or has
     been tampered with.
 
     **Note:** If your backup was compressed with a program other than `gzip`,
     you must substitute the correct compression program in the command above.
-    This information is contained in `backup-header` (see step 5). For example,
+    This information is contained in `backup-header` (see step 4). For example,
     if your backup is compressed with `bzip2`, use `bzip2 -d` instead of `gzip
     -d` in the command above.
 
- 8. Enter the decrypted directory, mount `private.img`, and access your data.
+ 7. Search inside of `qubes.xml` for the `backup-path` property of the qube
+    whose data you wish to restore. Using the value of this property (e.g.
```

Comment:
I think we need to provide instructions for this part (i.e., how exactly to "search inside of `qubes.xml` for the `backup-path` property) and probably break it up into separate steps.

I'm guessing it should be something like this:

```
$ grep 'backup-path' qubes.xml
        <feature name="backup-path">vm75/</feature>
```

In this test run, my backup contains only a single VM, which I'm guessing is why this returned only one result.

If the backup contains multiple qubes, the user will naturally wonder, at this point, how she's supposed to know which `vm##` corresponds to the qube she wants. For example, "Okay, what's the number for my `work` qube?"

Reply:
The *implied* instructions are something like

"Open `qubes.xml` in a text viewer. As XML is a famously readable format, you will intuitively grasp its overall structure and notice multiple `domain` elements (a domain being a VM, which is of course a qube). Find your qube `foo` by looking for a `domain` element containing a descendant element `<property name="name">foo</property>`. The `domain` element will also contain a descendant element\* `<feature name="backup-path">vm123/</feature>`. We will not insult you by writing down any part of this."

\* Omg it isn't even a "property" but a "feature" :facepalm:

> (I know this is a preexisting problem and wasn't introduced by your changes, but these new instructions make it more salient to me now for some reason.)

Besides making it more visible, the changes also made UX worse when the backup file only contains a single qube anyway. Previously it was possible to ignore `qubes.xml` in that case, and just go with the only extracted `vm123/` directory.

Ugh. Here's an (already unpleasantly long) command that produces a list of every VM in `qubes.xml`, plus the directory name for those that are included in the backup:

```
$ xmlstarlet sel -T -t -m //domain -v 'concat(.//property[@name="name"], " ", .//feature[@name="backup-path"])' -n qubes.xml
```

Reply:
> "[...]We will not insult you by writing down any part of this."

IMHO, that's the opposite attitude of the one we should take regarding these sorts of instructions, where the context is that the user has an emergency (by assumption, as these are *emergency* restore instructions) and is being walked through an unfamiliar procedure step-by-step, with no assumed skill level.

(It's kind of like the movie trope where a surgeon has to tell a panicked layperson how to perform surgery over the phone, or printing "in case of emergency, break glass" on a glass-encased object instead of assuming that people will figure it out in the moment. Better to over-explain even what seems obvious, just in case.)

Reply:
What, you don't agree that "draw the rest of the owl" is fine as a load-bearing step in data recovery, and that it would be gauche to assume that the reader hasn't already built a workable XML parser in their head? (I'm putting an `/s` here this time)

Latest commit has the `xmlstarlet` command.

Reply:
"Rest of the owl" is exactly what I thought of too. :joy:

Reply:
> Here's an (already unpleasantly long) command that depends on the `xmlstarlet` package and produces a list of every qube in `qubes.xml`, plus the directory name for those that are included in the backup: [...]

I like the neat and easily-readable output of this, but if we're concerned about the need for another uncommon package and a long command, what about something like this?

List all qubes included in the backup:

```
$ grep -B 14 'backup-path' qubes.xml | grep 'name="name"'
        <property name="name">work</property>
        <property name="name">untrusted</property>
        <property name="name">vault</property>
```

Find the QID for the one you want to restore:

```
$ grep -A 2 'vault' qubes.xml | grep 'qid'
        <property name="qid">75</property>
```

Then just add `vm` to the front and `/` to the end, so `vm75/`?

This way, we don't have to recommend that users store a copy of `xmlstarlet` with their backups just to be able to parse XML, and the commands are easier to type and understand. What do you think?

Reply:
Adding yet another dependency on xmlstarlet (plus its own dependencies) does suck. Although it seems to be more widely available through package managers than scrypt.

I'm not sure yet if it's possible to rely on the order and format in which XML elements are written out to `qubes.xml` across all versions of the code that generates v4 backups. That's always an annoying part of interacting with XML on a string level.

Currently the `grep` commands are also loose, like "vault" would match "anothervault" or "vault2" as well. I'd guess that at the point where the commands have been narrowed down to avoid most bycatch, they'll be even gnarlier than the `xmlstarlet` command line. In comparing the two approaches it might come down to avoiding the extra package dependency.

Reply:
Darn. :\

Maybe we should just provide the user with general instructions for using a text editor to visually read the XML and find what they need.

@marmarek, do you have any suggestions about this? (At the very least, this seems like something to keep in mind for any future design changes to the backup system.)

Reply:
Reposting my comment from [this discussion](https://forum.qubes-os.org/t/19395/):

> @rustybird: For what it's worth, if pushing the new instructions through even with the added `xmlstarlet` dependency is the lesser of two evils, I won't stand in the way. I prefer to avoid extra dependencies where feasible (to avoid exactly the "special software required"/"vendor lock-in" perception discussed above), but when you need 'em, you need 'em (like with `scrypt`).

Reply:
Yeah all in all `xmlstarlet` doesn't seem too bad to me.

I've added two more small commits incorporating some of the [feedback from @/tree in the forum](https://forum.qubes-os.org/t/getting-stuck-on-emergency-backup-recovery/18520/5).

How do you feel about setting aside other orthogonal improvements, like rewriting how to properly gather various package dependencies in advance (big can of worms...) and reformatting code blocks?

Reply:
That's also okay with me. I don't want to allow the perfect to be the enemy of the good.

Reply:
@marmarek This might have fallen through the cracks. Can you give us a yay or nay? I always feel weird linking people in the forum to my little branch instead of the official instructions.